ktor-network/posix/src/io/ktor/network/sockets/DatagramSocketImpl.kt

```diff
@@ -0,0 +1,145 @@
+    private val _localAddress: NetworkAddress,
+    private val _remoteAddress: NetworkAddress?,
+    parent: CoroutineContext = EmptyCoroutineContext
+) : BoundDatagramSocket, ConnectedDatagramSocket, ReadWriteSocket,
+    BaseNativeSocket(descriptor, selector, parent) {
+
+    override val localAddress: NetworkAddress
+        get() = _localAddress
+    override val remoteAddress: NetworkAddress
+        get() = _remoteAddress!! // TODO: What should happen here?
+
+    private val sender = BroadcastChannel<Datagram>(Channel.BUFFERED)
+
+    private val receiver = produce<Datagram>(coroutineContext) {
+        while (true) {
+            val received = receiveImpl()
+            channel.send(received)
+        }
+    }
+
+    init {
+        makeShared()
+
+        launch {
+            sender.consumeEach { datagram ->
+                sendImpl(datagram)
+            }
+        }
+    }
+
+    private tailrec suspend fun sendImpl(
+        datagram: Datagram,
+        bytes: ByteArray = datagram.packet.readBytes()
+    ) {
+        var result: Int? = null
+        bytes.usePinned { pinned ->
+            datagram.address.address.nativeAddress { address, addressSize ->
+                result = sendto(
+                    descriptor,
+                    pinned.addressOf(0),
+                    bytes.size.convert(),
+                    0,
+                    address,
+                    addressSize
+                ).toInt()
+            }
+        }
+        result!!
+
+        when (result!!) {
+            0 -> {
+                throw CancellationException("Socket was closed")
+            }
+            -1 -> {
+                if (errno == EAGAIN) {
+                    selector.select(selectable, SelectInterest.WRITE)
+                    sendImpl(datagram, bytes)
+                } else {
+                    throw PosixException.forErrno()
+                }
+            }
+            else -> return
+        }
+    }
+
+    private tailrec suspend fun receiveImpl(
+        buffer: IoBuffer = DefaultDatagramByteBufferPool.borrow()
+    ): Datagram {
+        memScoped {
+            val clientAddress = alloc<sockaddr_storage>()
+            val clientAddressLength: UIntVarOf<UInt> = alloc()
+            clientAddressLength.value = sockaddr_storage.size.convert()
+
+            val count = try {
+                buffer.write { memory, startIndex, endIndex ->
+                    val bufferStart = memory.pointer + startIndex
+                    val size = endIndex - startIndex
+                    val result = recvfrom(
+                        descriptor,
+                        bufferStart,
+                        size.convert(),
+                        0,
+                        clientAddress.ptr.reinterpret(),
+                        clientAddressLength.ptr
+                    ).toInt()
+
+                    if (result == 0) {
+                        throw CancellationException("Socket was closed")
```

Comment:
Should this be `CancellationException` or `IOException`?

Reply:
Should probably be `IOException` because that is used in the `attachForWriting` function as well.

Reply:
`IOException` for sure

Reply:
Ok, changed to IOException